[#7362] Add kafka client plugin IT

acafela opened the pull request:
issue : #7362 

This PR is kafka client plugin IT module.

KafkaUnitServer.java and ZookeeperUnitServer.java are the copy of https://github.com/chbatey/kafka-unit.
These enables Kafka broker, ZooKeeper to run in unit test.
KafkaUnitServer.java and ZookeeperUnitServer.java have been modified to suit the test environment.

koo-taejin commented:
hi @acafela 

Thank you for Your PRsending PR, It was really necessary for guaranteed Kafka Plugin stability. :) 

Please read following suggestion. 
#### 1. Issue of max time(current : 5000ms) in awaitTraceCount
I did executed it, and I found out that it failed sometimes in my environment.
In my environment, it sometimes took a time than time that you expected, so I suggest you to change max value(current : 5000ms) in awaitTraceCount to 10000ms.

#### 2. Issue of version that not yet released. 
I checked that you delcared all kafka client version in Test. 
In this case, if kafka-client is going to release for reason like backport. 
Kafka Plugin does not guaranteed stability of new version. 

If you is going to declare like following, Kafka Plugin is going to guaranteed stability of new version. 
```
// Ex) in KafkaClient_0_11_x_IT
        "org.apache.kafka:kafka-clients:[0.11.0.0],[0.11.0.1],[0.11.0.2],[0.11.0.3]",
        "org.apache.kafka:kafka-clients:[0.11.0.0,0.11.max]",
```

In the case of the latest version 2_6_x_IT, in order to automatically perform tests such as 2.7~, 2.8~ in the future release
If you specify as below, the test is automatically performed.
And it could make to check guaranteed compatibility of new version.
```
// Ex) in KafkaClient_2_6_x_IT
        "org.apache.kafka:kafka-clients:[2.6.0]"
        "org.apache.kafka:kafka-clients:[2.6.0,)"
```

thanks :)

acafela commented:
@koo-taejin Thank you for your review. I was able to understand and modify it because of your detailed explanation.

<br>

> #### 1. Issue of max time(current : 5000ms) in awaitTraceCount

I changed the await time to 10000ms as you suggested.

<br>

> #### 2. Issue of version that not yet released.

Thanks to your review, I got to know Aether Version Schema this time. I changed it so that I could test the new version.

<br>

As a check point, The kafka-client 2.3.1 and kafka broker 2.3.0 are not compatible, so i created two separate files: KafkaClient_2_3_0_IT, kafkaClient_2_3_1_to_max_IT.

koo-taejin commented on plugins-it/kafka-it/src/test/java/com/navercorp/pinpoint/plugin/kafka/KafkaClientITBase.java:
```diff
@@ -0,0 +1,142 @@
+
+    private static KafkaUnitServer kafkaUnitServer = new KafkaUnitServer(2189, 9092);
+    protected static final TestProducer producer = new TestProducer();
+    private static ExecutorService executorService = Executors.newSingleThreadExecutor();
```
Using daemon thread in Test could cause a bad situation. 
Please change this thread to daemon. 

If you need any reference, please check following code. 
`com.navercorp.test.pinpointplugin.activemqPollingMessageReceiver`
or change ThreadFactory that generates daemon thread.

acafela replied:
@koo-taejin
As you said, it is appropriate to be daemon thread.
I changed the polling thread to the daemon thread.

koo-taejin commented on plugins-it/kafka-it/src/test/java/test/pinpoint/plugin/kafka/KafkaITConstants.java:
```diff
@@ -0,0 +1,35 @@
+    public static final int PARTITION = 0;
+    public static final String TRACE_TYPE_RECORD = "RECORD";
+    public static final String TRACE_TYPE_MULTI_RECORDS = "MULTI_RECORDS";
+    public static final long MAX_TRACE_WAIT_TIME = 100000;
```
I think 100 sec is too high value. I think this value will be good about 10 sec to 30 sec.

acafela replied:
I made a typo😅.
It was modified to 10 seconds.

acafela commented:
Hi @koo-taejin
I modified the polling thread and awaitTimeout.
Please review it.

koo-taejin commented:
Hi @acafela 
I am going to check it 
Thank you for notifying. :)

koo-taejin commented:
@acafela 

Everything looks good.

Please squash all commits that you have sent to one commit, then I am going to merge it.

thanks :)

acafela commented:
@koo-taejin 

I made one commit.
Thank you for your review 🙏
